drivers/mdio/mdio_stm32_hal.c

```diff
@@ -73,27 +88,74 @@ static int mdio_stm32_write(const struct device *dev, uint8_t prtad,
 	return ret;
 }
 
-static void mdio_stm32_bus_enable(const struct device *dev)
-{
-	ARG_UNUSED(dev);
-}
-
-static void mdio_stm32_bus_disable(const struct device *dev)
-{
-	ARG_UNUSED(dev);
-}
-
 static int mdio_stm32_init(const struct device *dev)
 {
 	struct mdio_stm32_data *const dev_data = dev->data;
 	const struct mdio_stm32_config *const config = dev->config;
+	const struct device *clock = DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE);
+	ETH_HandleTypeDef *heth = &dev_data->heth;
 	int ret;
 
+	/* enable clock */
+	ret = clock_control_on(clock, (clock_control_subsys_t)&config->pclken);
+	if (ret < 0) {
+		LOG_ERR("Failed to enable ethernet clock needed for MDIO (%d)", ret);
+		return ret;
+	}
+
 	ret = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
 	if (ret < 0) {
 		return ret;
 	}
 
+#ifdef CONFIG_ETH_STM32_HAL_API_V2
+	HAL_ETH_SetMDIOClockRange(heth);
+#else /* CONFIG_ETH_STM32_HAL_API_V1 */
+	/* The legacy V1 HAL API does not provide a way to set the MDC clock range
+	 * via a separated function call.
+	 */
+	uint32_t tmpreg1 = 0U;
+	uint32_t hclk = 60000000U;
+	/* Get the ETHERNET MACMIIAR value */
+	tmpreg1 = (heth->Instance)->MACMIIAR;
+	/* Clear CSR Clock Range CR[2:0] bits */
+	tmpreg1 &= ETH_MACMIIAR_CR_MASK;
+
+	/* Get hclk frequency value */
+	hclk = HAL_RCC_GetHCLKFreq();
```

Comment:
Do you actually need hclk or bus freq that you can get with `clock_control_get_rate()` ?

Reply:
I copied this part straight from the hal, but that can be replaced with CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC